File: src/main/java/org/BeehiveRobotics/Library/Util/Java/BROpMode.java
package org.BeehiveRobotics.Library.Util.Java;

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode;

public abstract class BROpMode extends LinearOpMode {
    public enum OpModeType {
        Autonomous, TeleOp
    }

    private OpModeType opModeType = OpModeType.Autonomous;

    public abstract void initialize();

    public abstract void run();

    public abstract void end();

    protected final void setOpModeType(OpModeType opModeType) {
        this.opModeType = opModeType;
    }

    public final void addLine(String line) {
        telemetry.addLine(line);
        telemetry.update();
    }

    public final void addData(String title, String data) {
        telemetry.addData(title, data);
        telemetry.update();
    }

    protected Controller controller1;
    protected Controller controller2;

    public final void runOpMode() throws InterruptedException {
        try {
            controller1 = new Controller(gamepad1);
            controller2 = new Controller(gamepad2);
            initialize();
            waitForStart();
            switch (opModeType) {
                case TeleOp:
                    while (opModeIsActive()) {
                        run();
                        controller1.update();
                        controller2.update();
                    }
                case Autonomous:
                    run();
            }
        } finally {
            end();
        }
    }
}

File: src/main/kotlin/org/BeehiveRobotics/Library/Util/BROpMode.kt
package org.BeehiveRobotics.Library.Util

import com.qualcomm.robotcore.eventloop.opmode.LinearOpMode
import org.BeehiveRobotics.Library.Util.Controller

abstract class BROpMode(private val opModeType: OpModeType): LinearOpMode() {

    protected lateinit var controller1: Controller
    protected lateinit var controller2: Controller

    val dashboard = BRTelemetry(this)

    enum class OpModeType {
        Autonomous, TeleOp
    }

    abstract fun initialize()

    abstract fun run()

    open fun end() {}

    open fun onStartPressed() {}

    @Throws(InterruptedException::class)
    override fun runOpMode() {
        try {
            dashboard.addLine("Not ready to start")
            dashboard.update()
            controller1 = Controller(gamepad1)
            controller2 = Controller(gamepad2)
            initialize()
            dashboard.addLine("Ready to Start")
            dashboard.update()
            waitForStart()
            onStartPressed()
            when (opModeType) {
                BROpMode.OpModeType.TeleOp -> {
                    while (opModeIsActive()) {
                        run()
                        controller1.update()
                        controller2.update()
                    }
                }
                BROpMode.OpModeType.Autonomous -> run()
            }
            end()
        } finally {}
    }
}